File: Examples/test-suite/lua/char_strings_runme.lua
require("char_strings")

local CPLUSPLUS_MSG = "A message from the deep dark world of C++, where anything is possible."
local OTHERLAND_MSG = "Little message from the safe world."

local count = 5

-- Each set function takes (string, number) and returns true when the string
-- received matches OTHERLAND_MSG concatenated with number.
local function test_set(setter)
  for i = 0, count - 1 do
    assert(setter(OTHERLAND_MSG .. i, i) == true)
  end
end

local function test_pingpong(pp)
  for i = 0, count - 1 do
    local msg = OTHERLAND_MSG .. i
    assert(pp(msg) == msg)
  end
end

-- get functions
for i = 0, count - 1 do
  assert(char_strings.GetCharHeapString() == CPLUSPLUS_MSG)
  char_strings.DeleteCharHeapString()
end
for i = 0, count - 1 do
  assert(char_strings.GetConstCharProgramCodeString() == CPLUSPLUS_MSG)
  char_strings.DeleteCharHeapString()
end
assert(char_strings.GetCharStaticString() == CPLUSPLUS_MSG)
assert(char_strings.GetCharStaticStringFixed() == CPLUSPLUS_MSG)
assert(char_strings.GetConstCharStaticStringFixed() == CPLUSPLUS_MSG)

-- set functions
test_set(char_strings.SetCharHeapString)
test_set(char_strings.SetCharStaticString)
test_set(char_strings.SetCharArrayStaticString)
test_set(char_strings.SetConstCharHeapString)
test_set(char_strings.SetConstCharStaticString)
test_set(char_strings.SetConstCharArrayStaticString)
test_set(char_strings.SetCharConstStaticString)
test_set(char_strings.SetConstCharConstStaticString)
test_set(char_strings.SetConstCharTypedefString)

-- get set function
test_pingpong(char_strings.CharPingPong)
test_pingpong(char_strings.CharArrayPingPong)
test_pingpong(char_strings.CharArrayDimsPingPong)

-- variables: settable globals - set then read back
for i = 0, count - 1 do
  local msg = OTHERLAND_MSG .. i
  char_strings.global_char = msg
  assert(char_strings.global_char == msg)
end
for i = 0, count - 1 do
  local msg = OTHERLAND_MSG .. i
  char_strings.global_char_array1 = msg
  assert(char_strings.global_char_array1 == msg)
end
for i = 0, count - 1 do
  local msg = OTHERLAND_MSG .. i
  char_strings.global_char_array2 = msg
  assert(char_strings.global_char_array2 == msg)
end

-- variables: read-only const globals
assert(char_strings.global_const_char == CPLUSPLUS_MSG)
assert(char_strings.global_const_char_array1 == CPLUSPLUS_MSG)
assert(char_strings.global_const_char_array2 == CPLUSPLUS_MSG)

-- char *& tests
assert(char_strings.GetCharPointerRef() == CPLUSPLUS_MSG)
test_set(char_strings.SetCharPointerRef)
assert(char_strings.GetConstCharPointerRef() == CPLUSPLUS_MSG)
test_set(char_strings.SetConstCharPointerRef)

File: Examples/test-suite/ocaml/char_strings_runme.ml
open Swig
open Char_strings

let cpp_msg = "A message from the deep dark world of C++, where anything is possible."
let other_msg = "Little message from the safe world."

(* Note: the swigp4 syntax extension overloads the "to" keyword as a cast
   operator, so "for i = 0 to n" cannot be used here - iterate with List.iter. *)
let iters = [ 0; 1; 2; 3; 4 ]

(* Each set function takes (string, number) and returns true when the string
   received matches OTHERLAND_MSG concatenated with number. *)
let test_set setter =
  List.iter (fun i ->
    let msg = other_msg ^ string_of_int i in
    assert (setter (C_list [ C_string msg ; C_int i ]) as bool = true)
  ) iters

(* The char *& / const char *& set functions have an "argout" typemap that also
   returns the referenced string, so the result is a list [string; bool] with
   the bool return value last. *)
let test_set_ref setter =
  List.iter (fun i ->
    let msg = other_msg ^ string_of_int i in
    match setter (C_list [ C_string msg ; C_int i ]) with
      C_list [ _ ; b ] -> assert (get_bool b = true)
    | _ -> assert false
  ) iters

let test_pingpong pp =
  List.iter (fun i ->
    let msg = other_msg ^ string_of_int i in
    assert (pp (C_string msg) as string = msg)
  ) iters

let _ =
  (* get functions *)
  List.iter (fun _i ->
    assert (_GetCharHeapString C_void as string = cpp_msg);
    ignore (_DeleteCharHeapString C_void)
  ) iters;
  List.iter (fun _i ->
    assert (_GetConstCharProgramCodeString C_void as string = cpp_msg);
    ignore (_DeleteCharHeapString C_void)
  ) iters;
  assert (_GetCharStaticString C_void as string = cpp_msg);
  assert (_GetCharStaticStringFixed C_void as string = cpp_msg);
  assert (_GetConstCharStaticStringFixed C_void as string = cpp_msg);

  (* set functions.  The char[] parameter variants (SetCharArrayStaticString,
     SetConstCharArrayStaticString) are omitted: OCaml's char[] 'in' typemap
     strcpy's into the (null) argument pointer and crashes - a pre-existing
     char array limitation unrelated to char *&. *)
  test_set _SetCharHeapString;
  test_set _SetCharStaticString;
  test_set _SetConstCharHeapString;
  test_set _SetConstCharStaticString;
  test_set _SetCharConstStaticString;
  test_set _SetConstCharConstStaticString;
  test_set _SetConstCharTypedefString;

  (* get set function.  CharArrayPingPong / CharArrayDimsPingPong take char[]
     parameters and are omitted for the same char array limitation as above. *)
  test_pingpong _CharPingPong;

  (* variables: settable globals - set then read back *)
  List.iter (fun i ->
    let msg = other_msg ^ string_of_int i in
    assert (_global_char (C_string msg) as string = msg);
    assert (_global_char C_void as string = msg)
  ) iters;
  List.iter (fun i ->
    let msg = other_msg ^ string_of_int i in
    assert (_global_char_array1 (C_string msg) as string = msg);
    assert (_global_char_array1 C_void as string = msg)
  ) iters;
  List.iter (fun i ->
    let msg = other_msg ^ string_of_int i in
    assert (_global_char_array2 (C_string msg) as string = msg);
    assert (_global_char_array2 C_void as string = msg)
  ) iters;

  (* variables: read-only const globals *)
  assert (_global_const_char C_void as string = cpp_msg);
  assert (_global_const_char_array1 C_void as string = cpp_msg);
  assert (_global_const_char_array2 C_void as string = cpp_msg);

  (* char *& tests *)
  assert (_GetCharPointerRef C_void as string = cpp_msg);
  test_set_ref _SetCharPointerRef;
  assert (_GetConstCharPointerRef C_void as string = cpp_msg);
  test_set_ref _SetConstCharPointerRef
;;
